nautobot/core/nautobot_app/__init__.py

```diff
@@ -0,0 +1,169 @@
+from django.apps import AppConfig
+from django.urls import reverse
+from operator import getitem
+from collections import OrderedDict
+
+from nautobot.extras.choices import CustomLinkButtonClassChoices
+from nautobot.extras.plugins.utils import import_object
+from nautobot.extras.registry import registry
+
+
+class NautobotConfig(AppConfig):
+    menu_tabs = "navigation.menu_tabs"
+
+    def ready(self):
+        if self.__module__.endswith(".apps"):
+            self.__module__ = self.__module__[:-5]
+        menu_items = import_object(f"{self.__module__}.{self.menu_tabs}")
+        if menu_items is not None:
+            register_menu_items(menu_items)
+
+
+def register_menu_items(class_list):
+    for nav_tab in class_list:
+        registry["nav_menu"]["tabs"][nav_tab.name] = {
+            "weight": nav_tab.weight,
+            "groups": registry["nav_menu"]["tabs"].get(nav_tab.name, {}).get("groups", {}),
+            "permissions": [],
+        }
+
+        tab_perms = []
+        registry_groups = registry["nav_menu"]["tabs"][nav_tab.name]["groups"]
+        for group in nav_tab.groups:
+            registry_groups[group.name] = {
+                "weight": group.weight,
+                "items": registry_groups.get(group.name, {}).get("items", {}),
+                "permissions": [],
+            }
+
+            group_perms = []
+            for item in group.items:
+                registry_groups[group.name]["items"][item.link] = {
+                    "link_text": item.link_text,
+                    "weight": item.weight,
+                    "buttons": item.buttons,
+                    "permissions": item.permissions,
+                }
+                group_perms += item.permissions
+
+            registry_groups[group.name]["items"] = OrderedDict(
+                sorted(registry_groups[group.name]["items"].items(), key=lambda x: getitem(x[1], "weight"))
+            )
+            registry_groups[group.name]["permissions"] = group_perms
+            tab_perms += group_perms
+
+        registry["nav_menu"]["tabs"][nav_tab.name]["permissions"] += tab_perms
+        registry["nav_menu"]["tabs"][nav_tab.name]["groups"] = OrderedDict(
+            sorted(registry_groups.items(), key=lambda x: getitem(x[1], "weight"))
+        )
+
+    registry["nav_menu"]["tabs"] = OrderedDict(
+        sorted(registry["nav_menu"]["tabs"].items(), key=lambda x: getitem(x[1], "weight"))
+    )
+
+
+class NavMenuTab:
+    permissions = []
+    groups = []
+
+    def __init__(self, name, permissions=None, groups=None, weight=1000):
+        self.name = name
+        self.weight = weight
+        if permissions is not None:
+            if type(permissions) not in (list, tuple):
+                raise TypeError("Permissions must be passed as a tuple or list.")
+            self.permissions = permissions
+        if groups is not None:
+            if type(groups) not in (list, tuple):
```

Comment:
Same w/ the type check here as w/ `permissions`. Please use `isinstance()` instead.